File: sections/19_dijkstra_weighted_modeling/problems/c_discount_route/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m;
    if (!(cin >> n >> m)) return 0;
    vector<vector<pair<int,long long>>> adj(n);
    for (int i = 0; i < m; i++) {
        int u, v; long long w;
        cin >> u >> v >> w;
        adj[u - 1].push_back({v - 1, w});
    }
    const long long INF = (1LL << 62);
    vector<array<long long,2>> dist(n, {INF, INF});
    using State = tuple<long long,int,int>;
    priority_queue<State, vector<State>, greater<State>> pq;
    dist[0][0] = 0;
    pq.push({0, 0, 0});
    while (!pq.empty()) {
        auto [d, u, used] = pq.top(); pq.pop();
        if (d != dist[u][used]) continue;
        for (auto [v, w] : adj[u]) {
            long long nd = d + w;
            if (nd < dist[v][used]) {
                dist[v][used] = nd;
                pq.push({nd, v, used});
            }
            if (!used) {
                nd = d + w / 2;
                if (nd < dist[v][1]) {
                    dist[v][1] = nd;
                    pq.push({nd, v, 1});
                }
            }
        }
    }
    long long ans = min(dist[n - 1][0], dist[n - 1][1]);
    cout << (ans == INF ? -1 : ans) << '\n';
    return 0;
}

File: sections/19_dijkstra_weighted_modeling/problems/c_discount_route/solution.py
import heapq
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data); n = next(it); m = next(it)
    adj = [[] for _ in range(n)]
    for _ in range(m):
        u = next(it) - 1; v = next(it) - 1; w = next(it)
        adj[u].append((v, w))
    inf = 10**30
    dist = [[inf, inf] for _ in range(n)]
    dist[0][0] = 0
    heap = [(0, 0, 0)]
    while heap:
        d, u, used = heapq.heappop(heap)
        if d != dist[u][used]:
            continue
        for v, w in adj[u]:
            nd = d + w
            if nd < dist[v][used]:
                dist[v][used] = nd; heapq.heappush(heap, (nd, v, used))
            if used == 0:
                nd = d + w // 2
                if nd < dist[v][1]:
                    dist[v][1] = nd; heapq.heappush(heap, (nd, v, 1))
    ans = min(dist[n - 1])
    print(-1 if ans == inf else ans)


if __name__ == "__main__":
    main()

